### ISSUE TITLE
RSI constructor not consistent

### ISSUE THREAD

**Issue by Worthy7** [QuantConnect/Lean#1542]
The shorthand factory default is:
`public RelativeStrengthIndex RSI(Symbol symbol, TimeSpan resolution, int period, MovingAverageType movingAverageType = MovingAverageType.Simple, Func<IBaseData, decimal> selector = null)`

The constructor is:
`RelativeStrengthIndex(string name, int period, MovingAverageType movingAverageType = MovingAverageType.Wilders)`
        
https://en.wikipedia.org/wiki/Relative_strength_index says

> Some commercial packages, like AIQ, use a standard exponential moving average (EMA) as the average instead of Wilder's SMMA.

So I suppose the default for the shorthand needs changing to Wilders.

**Comment by jaredbroad** [comment]
Would be nice to find a second independent source to confirm wilders is industry default (ta lib source etc). Once confirmed - will need to review/run test cases to make sure they still pass and if not update the test.

**Comment by Worthy7** [comment]
I can confirm at least TradingView and XTick are using Wilders. Not sure about MetaTrader.

**Comment by jaredbroad** [comment]
Cool confirmed 👍 

http://svn.code.sf.net/p/ta-lib/code/branches/ta-lib/ta-lib/c/src/ta_func/ta_RSI.c

   /* Subsequent prevLoss and prevGain are smoothed
    * using the previous values (Wilder's approach).
    *  1) Multiply the previous by 'period-1'. 
    *  2) Add today value.
    *  3) Divide by 'period'.
    */
   prevLoss /= optInTimePeriod_0;
   prevGain /= optInTimePeriod_0;

   /* Often documentation present the RSI calculation as follow:
    *    RSI = 100 - (100 / 1 + (prevGain/prevLoss))
    *
    * The following is equivalent:
    *    RSI = 100 * (prevGain/(prevGain+prevLoss))
    *
    * The second equation is used here for speed optimization.
    */

**Comment by Worthy7** [comment]
It's a breaking change, so I think it needs handling the same way as the MACD issue too :/ message then change in the future.

**Comment by AlexCatarino** [comment]
In PR #1730 , we implement a moving average that applied to RSI (and ATR) reflects the way Welles Wilder computes this indicator. Unit tests were added. When/if it is merged, we can look into handling the changes you propose.

**Comment by Nucs** [comment]
@AlexCatarino, Any chance that the `RelativeStrenghtIndex` produces an inverted result?

Here is a quick comparison I've inducted:
I've received the following result with the pull request #1730:
![image](https://user-images.githubusercontent.com/649919/38274870-b885824e-3798-11e8-8162-ceb682d32050.png)
RSI seems to be inverted.
After changing the comparison operators in `RelativeStrengthIndex.cs` to the following code, it looked the way it should.
```C#
        protected override decimal ComputeNextValue(IndicatorDataPoint input)
        {
            if (previousInput != null && input.Value **<** previousInput.Value)
            {
                AverageGain.Update(input.Time, input.Value - previousInput.Value);
                AverageLoss.Update(input.Time, 0m);
            }
            else if (previousInput != null && input.Value **>=** previousInput.Value)
            {
                AverageGain.Update(input.Time, 0m);
                AverageLoss.Update(input.Time, previousInput.Value - input.Value);
            }
```

![image](https://user-images.githubusercontent.com/649919/38274992-1ee37046-3799-11e8-923f-e733736e8fda.png)

**Comment by AlexCatarino** [comment]
@Nucs, how did you produce the charts above?

**Comment by Nucs** [comment]
If you are refering to "visually how"; this is my custom desktop client with my own custom data.
Programmatically how:
```C#
//Initialize
_rsi = RSI(symbol, 14, MovingAverageType.Wilders, Resolution.Minute);
//OnData(TradeBar data)
Plot(_symbol, "RSI", _rsi);
```

**Comment by AlexCatarino** [comment]
Could you plug your data in the excel worksheet from [stockcharts](http://stockcharts.com/school/doku.php?id=chart_school:technical_indicators:relative_strength_index_rsi) and check out the results?
We have verified our indicator against external data and the part you are think is "inverted" hasn't been touched in the last patch.

**Comment by Nucs** [comment]
I concluded that it produces inverted result mainly by comparing the way it works on live charts or other websites that has it implemented.
* [TradingView](https://www.tradingview.com/wiki/Relative_Strength_Index_(RSI)) - theres a chart you can interact with.
* [Binance (Crypto)](binance.com/tradeDetail.html?symbol=BTC_USDT) Viewing any currency at 1 minute resolution while obviously enabling RSI will produce similar effect.
Google can help even more sources to verify that. those two were enough for me.

The chart you sent (stockchart's) does produce an inverted (atleast to my eye) rsi.
In fact, in one of my first days with LeanEngine I wrote my custom implementation to that specific xls chart you mentioned and it does too produce an inverted rsi.

Heres the code, (don't mind the bad coding - I didn't know much about the engine and just wanted it to run).
tl;dr: it is my own *old* implementation of RSI indicator to that chart.
``` C#

public class EbbyRSI : Indicator {
    public int Period { get; set; }
    public EbbyRSI(Symbol symbol, int period) : this($"EbbyRSI{period}_{symbol}",period) {}
    public EbbyRSI(string name, int period) : base(name) {
        if (name == null)
            throw new ArgumentNullException(nameof(name));
        if (period <= 1)
            throw new ArgumentOutOfRangeException(nameof(period),"Period must be atleast 2");

        this.Period = period;
        data = new List<IndicatorDataPoint>(period);
    }

    public override bool IsReady => _isReady;

    private decimal lastprice = 0;
    private decimal loss = 0;
    private decimal gain = 0;
    private DateTime At = DateTime.MinValue;
    private IndicatorDataPoint rs;
    private IndicatorDataPoint rsi = null;
    private bool _isReady = false;
    private List<IndicatorDataPoint> data;

    protected override decimal ComputeNextValue(IndicatorDataPoint input) {
        decimal change;
        var _loss = 0m;
        var _gain = 0m;
        At = input.Time;
        if (!_isReady) {
            data.Add(input);
            if (data.Count >= Period + 1) {
                change = data.Sum(d => d.Value) / Period;

                _isReady = true;
                decimal[] changes = new decimal[Period];
                changes[0] = 0m;

                for (int i = 1; i < Period + 1; i++) {
                    changes[i - 1] = data[i] - data[i - 1].Value;
                }

                gain = changes.Sum(d => d > 0 ? d : 0) / Period;
                loss = (changes.Sum(d => d < 0 ? d : 0) * -1) / Period;
                lastprice = input.Value;
                goto _rdy;
            }

            return 0m;
        }

        change = input - lastprice;
        lastprice = input;

        if (change == 0) {
            return rsi ?? 0m;
        }

        if (change > 0) {
            _gain = change;
        } else {
            _loss = change * -1;
        }

        gain = ((gain * (Period - 1)) + _gain) / Period;
        loss = ((loss * (Period - 1)) + _loss) / Period;

        _rdy:

        rs = new IndicatorDataPoint(At, gain / loss);
        rsi = new IndicatorDataPoint(At, _gain = (loss == 0 ? 100 : 100 - (100 / (1 + rs))));
        return _gain; //not really gain, just a temporary variable. returns RSI
    }
}
```

**Comment by AlexCatarino** [comment]
It would be great if you could get the data (bars and indicators) from third party like TradingView or Binance so that we could verify our implementation, because we cannot change the current one based on a visual evaluation.
At the moment, we have three models (ours, yours and stockchart's) that yields the same results and probably TradingView and Binance also do provided the same data.

**Comment by Worthy7** [comment]
@AlexCatarino  Please repoen, this is not resolved. @Nucs Don't issue hijack please

**Comment by Nucs** [comment]
@AlexCatarino Fair enough, Binance has a C# client which I can use to gather information in Lean's format and then compare visual results with their web client & hand-collected RSI values.
I'll go into depth the coming weekend - if it will be relevant, i'll open a new issue.
Cheers.
